libsoftwarecontainer/src/container.cpp

```diff
@@ -620,14 +627,40 @@ bool Container::bindMountInContainer(const std::string &pathInHost,
         if (!pathInList(tempPath)) {
             if (!touch(tempPath)) {
                 log_error() << "Could not create file " << tempPath;
+                m_create.rollBack();
                 return false;
             }
-            m_cleanupHandlers.push_back(new FileCleanUpHandler(tempPath));
         }
     }
 
-    return bindMountCore(pathInHost, pathInContainer, tempPath, readOnly);
+    if (!bindMountCore(pathInHost, pathInContainer, tempPath, readOnly)) {
+        // bindMountInContainer operation is not successful. Clean mounted tempPath.
+        MountCleanUpHandler rollback{tempPath};
+        rollback.clean();
+        // clean createdDirectories as well
+        m_create.rollBack();
+        return false;
+    }
 
+    // bindMountInContainer succeed. Add cleanup for mounted tempPath
+    if (pathIsDirectory) {
+        //Add clenuphandlers for created directories
+        if (!pathInList(tempPath)) {
+            m_cleanupHandlers.emplace_back(new DirectoryCleanUpHandler(tempPath));
+        }
+
+        // Add cleanup handlers for temporary directories
+        for (auto &tempFileCleaner : m_create.m_tempFileCleaners) {
+            m_cleanupHandlers.emplace_back(std::move(tempFileCleaner));
+        }
+        // clear rollback history
+        m_create.clear();
+    } else {
+        // add cleanup for tempPath file
+        m_cleanupHandlers.emplace_back(new FileCleanUpHandler(tempPath));
+    }
+    m_cleanupHandlers.emplace_back(new MountCleanUpHandler(tempPath));
+    return true;
```

Comment:
This feels largely like we're copying parts of the code from the FileToolKitWithUndo and making it more specific. Do we have to?